### ISSUE TITLE
Leaking file descriptor / socket within DomainSocket tooling?

### ISSUE THREAD

**Issue by knalli** [docker-java/docker-java#1956]
While searching for a resource leaking situation, I have found a possible leak issue in `DomainSocket` in case of a non existing resource.

This relevant component is a copy and the original one is from spring-boot. I had got in touch with them already, but in [their view respective in their use case they don't have a problem with this](https://github.com/spring-projects/spring-boot/issues/32423). However, this is still in issue for this project as the relevant resource is used in regular requests.

For example, an affected code call would be using the docker client with the apache client and a configured unix socket path. In that case, a unix/linux socket is opened.

As far as I can say, I would say that the call chain of 

```
# com.github.dockerjava.transport
DomainSocket.get(String path)
-> LinuxDomainSocket(String path)
-> DomainSocket(String path)
-> #open(String path)
```

results into a resource leak under the condition the `path` does not exist. I can see there is native code involved, because the exception happens in the connect method:

```
Caused by: com.sun.jna.LastErrorException: [2] No such file or directory
 at com.github.dockerjava.transport.LinuxDomainSocket.connect(Native Method)
 at com.github.dockerjava.transport.LinuxDomainSocket.connect(LinuxDomainSocket.java:49)
 at com.github.dockerjava.transport.DomainSocket.open(DomainSocket.java:69)
 at com.github.dockerjava.transport.DomainSocket.<init>(DomainSocket.java:59)
```

I have a demo provided at [this repository](https://github.com/knalli/poc-domainsocket-reasource-leaking/blob/main/src/test/java/com/example/demo/DemoTest.java).

```
// socket exist
BEGIN = Current open files: 70
Creating socket...
INIT = Current open files: 73
Closing socket..
EXIT = Current open files: 72

// socket does not exist
BEGIN = Current open files: 70
Creating socket...
Error: com.sun.jna.LastErrorException: [2] No such file or directory
INIT = Current open files: 73
No open socket
EXIT = Current open files: 73
```

Please have a look at the open files counter (which is based on the same as SB's FileDescriptorMetrics).

---

The symptom is the system (linux) reports for the java process with  `lsof -p <pid> | grep STREAM` multiple entries
```
java    1 user  240u     unix 0xffff9b1e9739f080       0t0 4063746316 type=STREAM
```

A couple of them are fine, but this list grows only in one of our applications. I took some time isolating the actual culprit, at the end speaking for this context the application does (reduced): call each minute `DomainSocket.get("/var/lib/docker.sock")`, but this socket does not exist on this system. (For sake of simplicity I have reduced the dynamic aspect here.)

On every "call", the list grows for more entries (which means more open files). The counter "open files" in the java runtime grows as well. The (linux) system reports this files as sockets without any meta information except the mapped pid.

---

I am not sure whether the file descriptor should stay (I would say no, because the exception is thrown) or the call must be guarded with "file exist" check. However, I guess this should be part of at least the DomainSocket class?

**Comment by neandrake** [comment]
I've just run into this as well, also using the apache http5 client. I wonder if the leak is related to the file descriptor being opened in the `DomainSocket` constructor rather than in the `connect(SocketAddress, int)` override from `AbstractSocket`. If errors occur during construction/connecting then the file descriptor might not get cleaned up properly, though I can't yet confirm that's the case. The other implementations of Socket used by `ApacheDockerHttpClientImpl` follow the lifecycle of establishing the connection within `connect()` rather than in the constructor.

**Comment by knalli** [comment]
For reference: It seems to be fixed by the Spring Boot team with v3.1.4, specifically with this patch https://github.com/spring-projects/spring-boot/commit/6805a3393b28c1a366f8685d632386464d7da7fc I have not checked out yet.

**Comment by neandrake** [comment]
The change/commit which @knalli referenced from the Sprint Boot project confirms what I've seen. It's a small change that try/catch the `connect()` to make sure `close()` is called if there are any errors when connecting.

I suspect a better fix would be to move the `connect()` out of the constructor and instead put it into the `connect(SocketAddress, int)` override.

Is there any means to re-open this task?
